File: acapella_api/app/Models/Track.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Track extends Model
{
    use SoftDeletes;

    protected $guarded = ['id'];

    protected $fillable = [
        'album_id',
        'choir_id',
        'title',
        'file_path',
        'cover_path',
        'duration_sec',
        'bitrate',
        'is_premium',
        'track_number',
    ];

    protected $appends = ['duration_label'];

    protected function casts(): array
    {
        return [
            'is_premium'   => 'boolean',
            'duration_sec' => 'integer',
            'bitrate'      => 'integer',
            'track_number' => 'integer',
        ];
    }

    // ── Accessors ─────────────────────────────────────────────────────────────

    /**
     * Human-readable duration, e.g. "4:32".
     * Appended automatically on serialisation.
     */
    public function getDurationLabelAttribute(): string
    {
        $m = intdiv($this->duration_sec, 60);
        $s = $this->duration_sec % 60;

        return sprintf('%d:%02d', $m, $s);
    }

    // ── Relationships ─────────────────────────────────────────────────────────

    public function album(): BelongsTo
    {
        return $this->belongsTo(Album::class);
    }

    public function choir(): BelongsTo
    {
        return $this->belongsTo(Choir::class);
    }

    public function likedByUsers(): BelongsToMany
    {
        return $this->belongsToMany(User::class, 'liked_tracks')
            ->withTimestamps();
    }
}

File: acapella_manager/lib/models/track.dart
import 'album.dart';
import 'choir.dart';

class Track {
  final int id;
  final int? albumId;
  final int choirId;
  final String title;
  final String? filePath; // Hidden from non-premium if track is premium!
  final String? coverPath;
  final int durationSec;
  final String? durationLabel; // Accessor from Laravel backend
  final int bitrate;
  final bool isPremium;
  final int? trackNumber;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  // Contextual attributes
  final bool? isLiked;

  // Relationships
  final Album? album;
  final Choir? choir;

  Track({
    required this.id,
    this.albumId,
    required this.choirId,
    required this.title,
    this.filePath,
    this.coverPath,
    this.durationSec = 0,
    this.durationLabel,
    this.bitrate = 128,
    this.isPremium = false,
    this.trackNumber,
    this.createdAt,
    this.updatedAt,
    this.isLiked,
    this.album,
    this.choir,
  });

  factory Track.fromJson(Map<String, dynamic> json) {
    return Track(
      id: json['id'],
      albumId: json['album_id'],
      choirId: json['choir_id'],
      title: json['title'],
      filePath: json['file_path'],
      coverPath: json['cover_path'],
      durationSec: json['duration_sec'] ?? 0,
      durationLabel: json['duration_label'],
      bitrate: json['bitrate'] ?? 128,
      isPremium: json['is_premium'] == 1 || json['is_premium'] == true,
      trackNumber: json['track_number'],
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
      isLiked: json['is_liked'] == 1 || json['is_liked'] == true,
      album: json['album'] != null ? Album.fromJson(json['album']) : null,
      choir: json['choir'] != null ? Choir.fromJson(json['choir']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'album_id': albumId,
      'choir_id': choirId,
      'title': title,
      'file_path': filePath,
      'cover_path': coverPath,
      'duration_sec': durationSec,
      'bitrate': bitrate,
      'is_premium': isPremium,
      'track_number': trackNumber,
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }
}
